hotp/ta/hotp_ta.c

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright (c) 2017, Linaro Limited
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+#include <hotp_ta.h>
+#include <string.h>
+#include <tee_internal_api_extensions.h>
+#include <tee_internal_api.h>
+
+/*
+ * This macro checks whether the return value is not equal to TEE_SUCCESS, if it
+ * is not equal, then it will set the return value to 'res' and then goto
+ * 'label'.
+ */
+#define CHECK_EXIT(label, res, ret_val) \
+	if (ret_val != TEE_SUCCESS) { \
+		EMSG("0x%08x", ret_val); \
+		res = ret_val; \
+		goto label; }
+
+/*
+ * This macro checks the expected parameters, if they are not the same, then it
+ * will set 'res' to TEE_ERROR_BAD_PARAMETERS and will goto label 'label'.
+ */
+#define CHECK_EXP_PARAM_EXIT(label, res, pt, ept) \
+	if (pt != ept) { \
+		EMSG("Expected: 0x%x, got: 0x%x", ept, pt); \
+		res = TEE_ERROR_BAD_PARAMETERS; \
+		goto label; }
+
+/* The size of a SHA1 hash in bytes. */
+#define SHA1_HASH_SIZE 20
+
+/* GP says that for HMAC SHA-1, max is 512 bits and min 80 bits. */
+#define MAX_KEY_SIZE 64 /* In bytes */
+#define MIN_KEY_SIZE 10 /* In bytes */
+
+/* Dynamic Binary Code 2 Modulo, which is 10^6 according to the spec. */
+#define DBC2_MODULO 1000000
+
+/*
+ * Currently this only supports a single key, in the future this could be
+ * updated to support multiple users, all with different unique keys (stored
+ * using secure storage).
+ */
+static uint8_t K[MAX_KEY_SIZE];
```

Comment:
Not that it matters much, but we're usually using only lower case letters in variables.

Reply:
Yes, but since this is an example I wanted to keep the variables as defined in the spec (and I think I've been almost consistent doing that besides the counter).